src/core/lib/promise/try_concurrently.h

```diff
@@ -0,0 +1,326 @@
+    return Pending{};
+  }
+  template <int kDoneBit>
+  void Destroy(uint8_t) {}
+
+  template <typename P>
+  FusedSet<P> With(P x) {
+    return FusedSet<P>(std::move(x));
+  }
+};
+
+template <typename Main, typename PreMain, typename PostMain>
+class TryConcurrently {
+ public:
+  TryConcurrently(Main main, PreMain pre_main, PostMain post_main)
+      : done_bits_(0),
+        pre_main_(std::move(pre_main)),
+        post_main_(std::move(post_main)) {
+    Construct(&main_, std::move(main));
+  }
+
+  TryConcurrently(const TryConcurrently&) = delete;
+  TryConcurrently& operator=(const TryConcurrently&) = delete;
+  TryConcurrently(TryConcurrently&& other) noexcept
+      : done_bits_(0),
+        pre_main_(std::move(other.pre_main_)),
+        post_main_(std::move(other.post_main_)) {
+    GPR_DEBUG_ASSERT(other.done_bits_ == 0);
+    other.done_bits_ = HelperBits();
+    Construct(&main_, std::move(other.main_));
+  }
+  TryConcurrently& operator=(TryConcurrently&& other) noexcept {
+    GPR_DEBUG_ASSERT(other.done_bits_ == 0);
+    done_bits_ = 0;
+    other.done_bits_ = HelperBits();
+    pre_main_ = std::move(other.pre_main_);
+    post_main_ = std::move(other.post_main_);
+    Construct(&main_, std::move(other.main_));
+    return *this;
+  }
+
+  ~TryConcurrently() {
+    if (done_bits_ & 1) {
+      Destruct(&result_);
+    } else {
+      Destruct(&main_);
+    }
+    pre_main_.template Destroy<1>(done_bits_);
+    post_main_.template Destroy<1 + PreMain::Size()>(done_bits_);
+  }
+
+  using Result =
+      typename PollTraits<decltype(std::declval<PromiseLike<Main>>()())>::Type;
+
+  Poll<Result> operator()() {
+    auto r = pre_main_.template Run<Result, 1>(done_bits_);
+    if (auto* status = absl::get_if<Result>(&r)) {
+      GPR_DEBUG_ASSERT(!IsStatusOk(*status));
+      return std::move(*status);
+    }
+    if ((done_bits_ & 1) == 0) {
+      auto p = main_();
+      if (auto* status = absl::get_if<kPollReadyIdx>(&p)) {
+        done_bits_ |= 1;
+        Destruct(&main_);
+        Construct(&result_, std::move(*status));
+      }
+    }
+    r = post_main_.template Run<Result, 1 + PreMain::Size()>(done_bits_);
+    if (auto* status = absl::get_if<Result>(&r)) {
+      GPR_DEBUG_ASSERT(!IsStatusOk(*status));
+      return std::move(*status);
+    }
+    if ((done_bits_ & NecessaryBits()) == NecessaryBits()) {
+      return std::move(result_);
+    }
+    return Pending{};
+  }
+
+  template <typename P>
+  auto NecessaryPush(P p);
+  template <typename P>
+  auto NecessaryPull(P p);
+  template <typename P>
+  auto Push(P p);
+  template <typename P>
+  auto Pull(P p);
+
+ private:
+  constexpr uint8_t NecessaryBits() {
+    return 1 | (PreMain::NecessaryBits() << 1) |
+           (PostMain::NecessaryBits() << (1 + PreMain::Size()));
+  }
+  constexpr uint8_t AllBits() {
+    return (1 << (1 + PreMain::Size() + PostMain::Size())) - 1;
+  }
+  constexpr uint8_t HelperBits() { return AllBits() ^ 1; }
```

Comment:
Could you add a comment on what the return value of this function represents ?

Reply:
done